peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+
+
+Security Implications
+=====================
+
+There are no known security consequences arising from this PEP.
+
+
+How to Teach This
+=================
+
+[How to teach users, new and experienced, how to apply the PEP to their work.]
+
+
+Open Issues
+===========
+
+Assignment in ``__new__``
+-------------------------
+
+Immutable classes like :class:`fractions.Fraction` often do not define ``__init__``;
+instead, they perform initialization in ``__new__`` or classmethods. The proposed
+feature won't be useful to them. 
+
+OTOH, allowing assignment within ``__new__`` (and/or classmethods) could open way
+to non-trivial bugs:
+
+.. code-block:: python
+
+    class Foo:
+        # fully initialized objects
+        object_cache: ReadOnly[ClassVar[dict[int, Self]]] = {}
+
+        foo: ReadOnly[int]
+
+        def __new__(cls, foo: int) -> Self:
+            if foo + 1 in cls.object_cache:
+                # this instance is already initialized
+                self = cls.object_cache[foo + 1]
+
+            else:
+                # this instance is not
+                self = super().__new__(cls)
+
+            # assignment to an object which has been initialized before,
+            # breaking the invariant a read-only attribute can be assigned to
+            # only during its initialization?
+            self.foo = foo
+
+            cls.object_cache[foo] = self
+            return self
+
+To my understanding, properly detecting this problem would require type checkers
+to keep track of the "level of initialization" of an object.
+
+This issue doesn't seem to impact ``__init__``, since it's rather uncommon to
+ever rebind ``self`` within it to any other object, and type checkers could
+flag the action as whole.
+
+
+Extending initialization
+------------------------
+
+Mechanisms such as :func:`dataclasses.__post_init__` or attrs' `initialization hooks <https://www.attrs.org/en/stable/init.html#hooking-yourself-into-initialization>`_
+augment initialization by providing a set of dunder hooks which will be called
+once during instance creation. The current rules would disallow assignment in those
+hooks. Specifying any single method in the pep isn't enough, as the naming and
+functionality differs between mechanisms (``__post_init__`` vs ``__attrs_post_init__``).
+
+``ReadOnly[ClassVar[...]]`` and ``__init_subclass__``
+-----------------------------------------------------
+
+Should this be allowed?
+
+.. code-block:: python
+
+    class URI:
+        protocol: ReadOnly[ClassVar[str]] = ""
+
+        def __init_subclass__(cls, protocol: str = "") -> None:
+            cls.foo = protocol
+
+    class File(URI, protocol="file"): ...
+
+``Final`` in protocols
+----------------------
+
+It's been `suggested <https://discuss.python.org/t/expanding-readonly-to-normal-classes-protocols/67359/45>`_
+to clarify in this PEP whether ``Final`` should be supported by protocols.
```

Comment:
It would be good to clear this up, but it seems out of scope for this PEP, and like something that could be handled with a PR to the spec and approval by the typing council; it doesn't need a PEP.